FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/FreeRTOS_DNS.c

```diff
@@ -720,45 +1070,65 @@ TickType_t uxWriteTimeOut_ticks = ipconfigDNS_SEND_BLOCK_TIME_TICKS;
 			/* Get a buffer.  This uses a maximum delay, but the delay will be
 			capped to ipconfigUDP_MAX_SEND_BLOCK_TIME_TICKS so the return value
 			still needs to be tested. */
+			FreeRTOS_printf( ( "DNS lookup: \"%s\"\n", pcHostName ) );
 
 			uxHeaderBytes = ipSIZE_OF_ETH_HEADER + ipSIZE_OF_IPv4_HEADER + ipSIZE_OF_UDP_HEADER;
 
 			pxNetworkBuffer = pxGetNetworkBufferWithDescriptor( uxHeaderBytes + uxExpectedPayloadLength, 0UL );
 			if( pxNetworkBuffer != NULL )
 			{
 				pucUDPPayloadBuffer = &( pxNetworkBuffer->pucEthernetBuffer[ uxHeaderBytes ] );
+				/* Later when translating form UDP payload to a Network Buffer,
+				it is important to know whether this is an IPv4 packet. */
+				pucUDPPayloadBuffer[ - ( ( int ) ipUDP_PAYLOAD_IP_TYPE_OFFSET ) ] = ( uint8_t ) ipTYPE_IPv4;
 			}
 
 			if( pucUDPPayloadBuffer != NULL )
 			{
 				/* Create the message in the obtained buffer. */
-				uxPayloadLength = prvCreateDNSMessage( pucUDPPayloadBuffer, pcHostName, uxIdentifier );
+				uxPayloadLength = prvCreateDNSMessage( pucUDPPayloadBuffer, pcHostName, uxIdentifier, uxHostType );
 
 				iptraceSENDING_DNS_REQUEST();
 
-				/* Obtain the DNS server address. */
-				FreeRTOS_GetAddressConfiguration( NULL, NULL, NULL, &ulIPAddress );
-
 				/* Send the DNS message. */
 #if( ipconfigUSE_LLMNR == 1 )
 				if( bHasDot == pdFALSE )
 				{
 					/* Use LLMNR addressing. */
-					( ipCAST_PTR_TO_TYPE_PTR( DNSMessage_t, pucUDPPayloadBuffer ) )->usFlags = 0;
-					xAddress.sin_addr = ipLLMNR_IP_ADDR; /* Is in network byte order. */
+					( ipPOINTER_CAST( DNSMessage_t *, pucUDPPayloadBuffer ) ) -> usFlags = 0;
+					xAddress.sin_addr = ipLLMNR_IP_ADDR;	/* Is in network byte order. */
 					xAddress.sin_port = ipLLMNR_PORT;
 					xAddress.sin_port = FreeRTOS_ntohs( xAddress.sin_port );
 				}
 				else
 #endif
 				{
+				NetworkEndPoint_t *pxEndPoint;
+
+					ulIPAddress = 0UL;
+					/* Obtain the DNS server address. */
+					for( pxEndPoint = FreeRTOS_FirstEndPoint( NULL );
+						 pxEndPoint != NULL;
+						 pxEndPoint = FreeRTOS_NextEndPoint( NULL, pxEndPoint ) )
+					{
+						if( ENDPOINT_IS_IPv4( pxEndPoint ) )
+						{
+							FreeRTOS_GetAddressConfiguration( NULL, NULL, NULL, &ulIPAddress, pxEndPoint );
+							if( ( ulIPAddress != ( uint32_t )0UL ) && ( ulIPAddress != ~( uint32_t )0UL ) )
+							{
+								break;
+							}
+						}
+					}
 					/* Use DNS server. */
+					xAddress.sin_len = ( uint8_t ) sizeof( xAddress );
+					xAddress.sin_family = FREERTOS_AF_INET;
 					xAddress.sin_addr = ulIPAddress;
 					xAddress.sin_port = dnsDNS_PORT;
 				}
 
 				ulIPAddress = 0UL;
-
+/*logging*/FreeRTOS_printf( ( "Send message to %lxip\n", FreeRTOS_htonl( xAddress.sin_addr ) ) );
```

Comment:
What is this `/*logging*/` comment for?

Reply:
That was for me, to remember take away all lines that start with `/*logging*/`. But for now the logging was still interesting enough.